core/src/main/scala/tapir/Codec.scala

```diff
@@ -331,3 +365,20 @@ trait Decode[F, T] {
     }
   }
 }
+
+trait Validate[F, T] extends Decode[F, T] {
+
+  private[tapir] val validator: Validator[T] = Validator.passing
+
+  def decodeAndValidate(f: F): DecodeResult[T] = {
```

Comment:
now we have three methods: `decode`, `safeDecode` and `decodeAndValidate`. This might be confusing. Maybe we could rename `decode` to `rawDecode` or sth like that, and keep only `decode` which would (1) validate (2) catch exceptions (3) call `rawDecode`? That would also collapse the `Validate` and `Decode` traits into one.

I don't think there's any use-case in calling these methods separately

Reply:
Agree, but I at the same time I like the separation of `decode` and `validate`. Also `decode`/`encode` creates nice couple. 
How about renaming `decodeAndValidate` to `safeDecode`, leaving `Decode::safeDecode` (will be called using `super.safeDecode`) and also leaving `Codec::decode` without change?

Reply:
Separation as private methods: sure. But no need to introduce public traits/methods, which can only lead to confusion.

For me `safeDecode` doesn't say anything about validation, only exceptions. So I'd either go with `decode` - meaning the whole process (raw decode, exceptions, validations), or sth with "validation" in the name